exercises/04-rce-reverse-shell.md

```
@@ -0,0 +1,110 @@
+  <summary>Answer</summary>
+
+  This executes the `cowsay` program on our host linux box, but as you can see, it allows user input. This is a sign that it might have a RCE type vulnerability if input is not validated. Let's try with a few inputs to see what we have control over.
+
+  Let's assume that the executable looks something like `cowsay 'input'`. There wouldn't have to be a single quote there, but it's a reasonable initial assumption. We'll have to escape using the same method we used for the SQL injection. So let's see what we can execute the `ls` command to list the directory contents.
+
+  The way you'd escape would be something like this where the `';` gets you out of the expected command and then you execute your own command. Then the `#` comment character ignores the rest of the command just like in the SQL injection:
+
+  ```
+  cowsay 'input'; ls #
+  ```
+  But there's some characters that probably won't work in a URL parameter. This is where we'll try to encode our command string with a URL encoder. We'll use https://www.urlencoder.org/ for this. If you copy the above line and encode it you can add it to the URL parameter like so and then execute the curl request to see what you get:
```

Markdownlint

Bare URL used

(MD034, no-bare-urls)